scripts/jsondeps.gradle

```diff
@@ -21,31 +37,79 @@
+                println "DEBUG: Resolved dep"
+                println resolvedDep
+
+                println "DEBUG: Module artifacts"
+                println resolvedDep.moduleArtifacts
+                println resolvedDep.moduleArtifacts.size()
+
+                // moduleArtifacts never returns null, but sometimes this
+                // iterator can be empty (for dependencies with no artifacts
+                // e.g. `jackson-bom`).
+                //
+                // artifact: https://docs.gradle.org/current/javadoc/org/gradle/api/artifacts/ResolvedArtifact.html
+                def artifact = resolvedDep.moduleArtifacts.iterator().next()
+
+                println "DEBUG: Artifact"
+                println artifact
+
+                // artifact.id: https://docs.gradle.org/current/javadoc/org/gradle/api/artifacts/component/ComponentArtifactIdentifier.html
+                // artifact.id.componentIdentifier: https://docs.gradle.org/current/javadoc/org/gradle/api/artifacts/component/ComponentIdentifier.html
                 def id = artifact.id.componentIdentifier
                 def json = "{"
+
+                // A "project component identifier" is a dependency that refers
+                // to another Gradle subproject (as opposed to a third-party
+                // dependency). We don't care about these for now.
                 if (id instanceof ProjectComponentIdentifier) {
                     // Minor problem here: we don't get the specific configuration used for the subproject.
                     // The default is the configuration named "default"
-                    json += "\"type\":\"project\",\"name\":\"${id.projectPath}\"" 
+                    json += "\"type\":\"project\",\"name\":\"${id.projectPath}\""
                 } else if (id instanceof ModuleComponentIdentifier) {
+                    // A "module" is a third-party dependency. Almost all
+                    // modules have "artifacts", which is the actual dependency
+                    // code that gets downloaded.
                     json += "\"type\":\"package\",\"name\":\"${id.group}:${id.module}\",\"version\":\"${id.version}\","
                     def childResults = []
                     if (!resolvedDep.children.isEmpty()) {
+                        // childResolvedDep: https://docs.gradle.org/current/javadoc/org/gradle/api/artifacts/ResolvedDependency.html
                         resolvedDep.children.each { childResolvedDep ->
-                            def result = depToJSON childResolvedDep
-                            childResults << result
+                            // If we don't do this check, then calling
+                            // `resolvedDep.moduleArtifacts.iterator().next()`
+                            // above explodes because the `moduleArtifacts`
+                            // iterator is empty.
```

Comment:
It would be nice to move this check closer to the invocation of `iterator().next()` -- seems error-prone to keep them separate. Could return `null` from the recursive `depToJSON` call and do a null-check below before adding to `childResults`

Reply:
Good call, not sure what I was thinking here.